zcobol/test/YOUZPI2.CBL

```diff
@@ -0,0 +1,47 @@
+       IDENTIFICATION DIVISION.
+       PROGRAM-ID.    YOUZPI2.
+       AUTHOR.        JAMES FRANCIS CRAY.
+      *
+      * THIS PROGRAM CALCULATES YOUZPI FOR YOU
+      * YOUZPI FROM 4(1 - 1/3 + 1/5 - 1/7 …ETC.)
+      *
+      * 2022-03-17 Abe Kornelis: this program incurs S0C7 when run
+      *                          Quite likely a compiler bug
+      *                          No time now to investigate & fix
+      *
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+       01  COUNTER        PIC 9(05)      VALUE 1.
+       01  ITERATIONS     PIC 9(05)      VALUE 9999.
+       01  TERM           PIC S9999V9999 VALUE +0.
+       01  TEMP1          PIC S9999V9999 VALUE +0.
+       01  TEMP2          PIC S9999V9999 VALUE +0.
+       01  TEMP3          PIC S9999V9999 VALUE +0.
+       01  TEMP4          PIC S9999V9999 VALUE +0.
+       01  FAC            PIC S9         VALUE 1.
+       01  FAC1           PIC S9         VALUE 0.
+       01  MYSUM          PIC S9999V9999 VALUE +0.
+       01  MYSUM1         PIC S9999V9999 VALUE +0.
+       01  THANKS         PIC X(30)      VALUE
+           'YOU THANKS FOR WATCHING YOUR YOUZPI '.
+       
+       PROCEDURE DIVISION.
+       BEGIN.
+           DISPLAY 'YOU HELLO YOUR YOUZPI IS CALCULATING'.
+           PERFORM MYLOOP WITH TEST AFTER
+               VARYING COUNTER FROM 1 BY 1
+               UNTIL COUNTER GREATER THAN ITERATIONS.
+           MULTIPLY MYSUM BY 4.00000 GIVING MYSUM1.
+           DISPLAY ' YOUR YOU YOUZPI CALCULATED TO BE:'.
+           DISPLAY MYSUM1.
+           DISPLAY THANKS.
+           STOP RUN.
+
+       MYLOOP.
+           MULTIPLY COUNTER BY 2.0 GIVING TEMP1.
```

Comment:
Since this program is not part of a test, the following comments are documenting the S0C7 that occurs when attempting to run it.

This MULTIPLY is causing the S0C7. Reason follows.

Looking in YOU2PI2.PRN you will see that this MULTIPLY becomes
```
FC4390A8E000     MP    ZCVT_PWORK1(5),0(4,ZC_R14)
```
The value in ZCVT_PWORK1(5) is X'000010000C', which is 1, adjusted for the PICTURE used for TEMP1. The value in 0(4,ZC_R14) is X'0000020C', which is 2.0 (2 adjusted for one decimal place). The MP instruction will S0C7 if the first operand does not have 4 leading bytes of zeros -- the length of the second operand. As you can see, the first operand only has 2 leading bytes of zeros.

This is certainly a bug in the zCOBOL compiler. The MP instruction can get a specification exception (S0C6) if the multiplier (2nd operand) has length greater than 7 or the length of the multiplier exceeds the length of the multiplicand. It was easy to create test cases for each of these possibilities. For example, change the MULTIPLY instruction to "MULTIPLY COUNTER BY 2 GIVING TEMP1." 

YOUZPI2.CBL is the same as YOUZPI1.CBL except all numeric fields are zoned decimal -- an inefficient format when doing arithmetic calculations. For example, COUNTER is initialized to 1, which is X'F0F0F0F0F0'; FAC is initialized to 1 --- since it is signed, its value is X'C0'.

Lot of work for someone to do here.

Reply:
I have converted your analysis into a new issue #376. Thanks for your detailed analysis!